File: 2000290120152/week2/Stack_and_Queue.py
#SHIVANSHU SINGH 2000290120152
#1. STACK IMPLEMENTATION

class Stack:
    def __init__(self, size):
        self.stack = list()
        self.maxSize = size
        self.top = 0
        
    def push(self,data):
        if len(self.stack)>=self.maxSize:
            print("Stack Overflow!!")
            return 
        self.stack.append(data)
        self.top += 1
        
    def pop(self):
        if len(self.stack)==0:
            return ("Stack is Empty!!")
        item = self.stack.pop()
        self.top -= 1
        return item
        
    def display(self):
        if len(self.stack)==0:
            print("Empty Stack!!")
            return
        print(self.stack)
        

s = Stack(4)

s.push(1)
s.push(2)
s.push(3)
s.push(4)
s.push(5)
s.display()
print(s.pop()) 
print(s.pop()) 
print(s.pop()) 
print(s.pop()) 
print(s.pop()) 
print(s.pop()) 
s.display()

#--------------------------------2. QUEUE IMPLEMENTATION--------------

class Queue:
    def __init__(self):
        self.items = []
 
    def is_empty(self):
        return self.items == []
 
    def enqueue(self, data):
        self.items.append(data)
 
    def dequeue(self):
        return self.items.pop(0)
    
    def display(self):
        print(self.items)
        
        
        
q = Queue()

q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.enqueue(4)
q.enqueue(5)

q.display()

q.dequeue()
q.dequeue()
q.dequeue()

q.display()

print(q.is_empty())

File: 2000290120140/Stack and Queue.cpp
//Stack Implementation
#include<iostream>
using namespace std;
#define MAX 1000 
  
class Stack
{
   int top;
   public:
   int myStack[MAX]; 
  
   Stack() { top = -1; }
   bool push(int x);
   int pop();
   bool isEmpty();
};
   bool Stack::push(int item)
   {
      if (top >= (MAX-1)) {
      cout << "Stack Overflow!!!";
      return false;
   }
else {
   myStack[++top] = item;
   cout<<item<<endl;
   return true;
   }
}
int Stack::pop()
{
   if (top < 0) {
      cout << "Stack Underflow!!";
      return 0;
   }
else {
       int item = myStack[top--];
      return item;
   }
}
bool Stack::isEmpty()
{
   return (top < 0);
}
int main()
{
   class Stack stack;
   cout<<"The Stack Push "<<endl;
   stack.push(5);
   stack.push(7);
   stack.push(8);
   cout<<"The Stack Pop : "<<endl;
   while(!stack.isEmpty())
      {
      cout<<stack.pop()<<endl;
      }
   return 0;
}

//Queue Implemantation
#include <iostream>
using namespace std;
int queue[50];
int n = 50;
int front = - 1;
int rear = - 1;
void Queue_insertion() {
   int val;
   if (rear == n - 1)
      cout<<"Queue Overflow"<<endl;
   else {
    
      front = 0;
      cout<<" insert value in the queue : "<<endl;
      cin>>val;
      rear++;
      queue[rear] = val;
   }
}
void Delete() {
   if (front == - 1) {
      cout<<"Queue Underflow ";
   return ;
   } else {
      cout<<"Element deleted from queue is : "<< queue[front] <<endl;
      front++;;
   }
}
void Display_Queue () {
   if (front == - 1 )
   cout<<"Queue is empty"<<endl;
   else {
      cout<<"Queue elements are : ";
      for (int i = front; i <= rear; i++)
         cout<<queue[i]<<" ";
      cout<<endl;
   }
}
int main() {
   int ch;
   cout<<"1) insertion element to the queue"<<endl;
   cout<<"2) Delete element from queue"<<endl;
   cout<<"3) Display all the elements of queue"<<endl;
   cout<<"4) Exit"<<endl;
do {
   cout<<"Enter your choice : "<<endl;
   cin>>ch;
   switch (ch) {
      case 1: Queue_insertion();
         break;
      case 2: Delete();
         break;
      case 3: Display_Queue ();
         break;
      case 4: cout<<"Exit"<<endl;
         break;
      default: cout<<"Invalid choice"<<endl;
   }
} while(ch!=4);
   return 0;
}
